src/test/ui/const-generics/issues/issue-62456.stderr

```diff
@@ -6,3 +6,13 @@ LL | #![feature(const_generics)]
    |
    = note: `#[warn(incomplete_features)]` on by default
 
+error: constant expression depends on a generic parameter
+  --> $DIR/issue-62456.rs:5:20
+   |
+LL |     let _ = [0u64; N + 1];
+   |                    ^^^^^
+   |
+   = note: this may fail depending on what value the parameter takes
```

Comment:
I think we should improve the wording of the error. In my eyes I would prefer something along the lines of

```
error: constant expression can't depend on generic parameter
  --> $DIR/issue-62456.rs:5:20
   |
LL |     let _ = [0u64; N + 1];
   |                    ^^^^^ depends on a generic parameter
   |
   = note: this will be supported in the future, but isn't now
```

preferably with a link to documentation, tracking issue or its own error code so that we explain _why_ we don't support it yet and when it might be done.

Reply:
The error is indeed misleading, because what you're suggesting isn't what the error means.

The error is trying to say that *evaluating* the constant expression may only succeed for some "values" of the generic parameter.
The missing piece of the puzzle is a way to put that expression in a `where` clause to force the caller to evaluate it, ruling out post-monomorphization errors (#68436).

Note that [expressions using generics can still evaluate](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ba0ae8e0eb712b342a61a776ac52c62f) just fine while remaining polymorphic.

cc @varkor

Reply:
I agree that this diagnostic could be improved, especially as it was misinterpreted. How about something like:
```
error: constant expression may not depend on a generic parameter
  --> $DIR/issue-62456.rs:5:20
   |
LL |     let _ = [0u64; N + 1];
   |                    ^^^^^
   |
   = note: this expression may only be valid for some values of `N`, which `N` is guaranteed to take
```
This is a bit long, though this sort of thing hopefully gets the intention across better.

Reply:
"constant expression may **not** depend on a generic parameter"? (emphasis mine)

Reply:
```
   = note: this will be supported in the future, but isn't now
```
(Let's not make promises. ^^)

> "constant expression may **not** depend on a generic parameter"? (emphasis mine)

I think we tend to prefer **cannot* over **may not**. The former is more definitive.

Reply:
Maybe `is not allowed to depend`, which makes clear whether the constant expression is supposed to or not.

Reply:
But it *can* depend on generic parameters, the only problem is we have no *guarantor* for it *evaluating* successfully.

So perhaps a better phrasing for the error itself might be "cannot prove/determine/guarantee/etc. that constant expression will evaluate successfully" and then we can search its `Substs` for parameters and list them out to give the more helpful information.

We should also link [`https://github.com/rust-lang/rust/issues/68436`](https://github.com/rust-lang/rust/issues/68436), in some sort of help messaging along the lines of "there is no way to currently write the necessary `where` clause, watch this space".

Reply:
Well, I suppose I meant that it could not depend on an "unrestricted" generic parameter (or "guaranteed well-formed"). Yes, it's quite difficult to use language that's both intuitive, and doesn't take up a lot of space. Linking to the issue would be good.

Reply:
+1 on linking to that issue. I would say I've come around to the position that reducing verbosity is a goal, but much lower than correct understandable wording.

Reply:
Opened #71142 for this subthread.